perf(sql): add JIT compilation support for string/binary null check

mtopolnik opened the pull request:
Closes #4143.

The PR adds JIT compilation support for string/binary column null check. In the IR for the JIT compiler it models a null check as the equality of an operand of type "string_header" or "binary_header", and another operand of type i32 or i64, which is a constant -1.

The JIT compiler introduces an optimization for the case of mostly-nonempty values: instead of loading the header through a data-dependent load, it compares the two neighboring entries in the varlen index column. Their difference is equal to the total length of an entry in the data column. If this is more than `header_size`, the value is definitely present and there's no need to check it. If it equals `header_size`, the entry is either null or an empty value. To decide which, we must load the header from the varlen column.

For the purpose of AVX-2 mode, the size of the NULL check expression is 8 bytes for both a string and binary column. This is dictated by the 8-byte size of the offset in the varlen index column.

### Benchmark

I created a JMH benchmark `StringNullCheckBenchmark` which tests the following array of combinations:

1. Dataset statistics: 99% non-null, 67% non-null, 1% non-null
2. Mode: SIMD, Scalar, and JIT disabled.

It uses this DDL and query:

```sql
create table x as (select 
    rnd_str(100, 70, 140, nullFreq) string_value, 
    timestamp_sequence(to_timestamp('2024-02-04', 'yyyy-MM-dd'), 100000L) ts
  from long_sequence(10_000_000)) timestamp(ts);

select count(*) from x where string_value is null
```

Here `nullFreq` means "one in this many rows has a NULL value".

I got this result:

```
(nullFreq)  (jitMode)  Mode  Cnt    Score    Error  Units
         1       SIMD  avgt    5   53.273 ±  0.549  ms/op
         1     SCALAR  avgt    5   54.436 ±  0.841  ms/op
         1   DISABLED  avgt    5   80.439 ±  0.771  ms/op

         3       SIMD  avgt    5  131.997 ±  2.924  ms/op
         3     SCALAR  avgt    5  163.855 ± 11.379  ms/op
         3   DISABLED  avgt    5  461.568 ± 42.395  ms/op

       100       SIMD  avgt    5   22.734 ±  0.660  ms/op
       100     SCALAR  avgt    5   26.763 ±  0.857  ms/op
       100   DISABLED  avgt    5  454.270 ± 29.456  ms/op
```

The JIT optimization substantially improves query performance in all cases, and especially in the 99% non-null case (nullFreq = 100), where it gives a 20x advantage. Even when all rows have NULL, which triggers the slow path, the JIT-compiled version still gives a 50% advantage.

### The Worst Case

The worst case I could find for the JIT-compiled query is this: the filter is `string_value is null` and the dataset contains a mix of empty and short non-empty strings, with no nulls at all. Specifically, in my benchmark code I replaced

```
"rnd_str(100, 70, 140, " + a_nullFreq + ") string_value"
```

with

```
"rnd_str('aa', 'a', '') string_value"
```

In this particular case, this is the result:

```
(jitMode)  Mode  Cnt   Score   Error  Units
     SIMD  avgt    5  17.599 ± 0.534  ms/op
   SCALAR  avgt    5  36.614 ± 6.574  ms/op
 DISABLED  avgt    5   4.448 ± 0.222  ms/op
```

My interpretation of the reason is that the empty value triggers the slow path in JIT-compiled code, but the outcome is always the same as if the fast path was taken. Java code takes the slow path directly, and in this case there are no further overheads from writing out the row id (since the filter function always returns false), and the string data itself is short and well-packed into cache lines.

### Comparison of SIMD and Scalar Code

In the above results, SIMD code confers only a modest advantage, so I investigated another case: a more complex query that involves two more Long columns. I used this query:

```sql
select count(*) from x where long_value1 > 0 and long_value2 > 0 and string_value is null
```

This gave me the following results:

```
(nullFreq)  (jitMode)  Mode  Cnt    Score    Error  Units
         0       SIMD  avgt    5   21.142 ±  0.713  ms/op
         0     SCALAR  avgt    5   49.493 ±  2.678  ms/op
         0   DISABLED  avgt    5   15.038 ±  6.290  ms/op

         1       SIMD  avgt    5   88.372 ±  8.755  ms/op
         1     SCALAR  avgt    5   96.053 ±  3.497  ms/op
         1   DISABLED  avgt    5  148.575 ±  3.892  ms/op

         3       SIMD  avgt    5  215.233 ±  3.199  ms/op
         3     SCALAR  avgt    5  305.310 ± 21.980  ms/op
         3   DISABLED  avgt    5  693.793 ± 31.691  ms/op

       100       SIMD  avgt    5   31.662 ±  0.655  ms/op
       100     SCALAR  avgt    5   44.060 ±  2.040  ms/op
       100   DISABLED  avgt    5  679.549 ± 27.248  ms/op
```

We can see that, in the case with 33% and 1% nulls, the advantage of SIMD mode is quite significant, up to 50%. In the worst case, where Java code performs best, SIMD has the largest advantage over scalar and covers most of the gap with Java. This may serve as justification for the complexity the SIMD code adds to the codebase.

### Possible Improvements

The weak spot of the current SIMD implementation is that it only works when the whole query is on data of size 8. We could improve it to work with other sizes as well.

### Future Work

This work can be easily reused for the equality of string contents, especially if we want to bias the performance towards unequal strings. For that case we'd emit IR code of the shape `len(left) = len(right) AND left = right`. The code in this PR already does the length comparison, but it's currently blocked from checking any other case than equality with the constant -1.

I think it would be especially easy to add equality of two columns, although it's probably a less useful construct. To allow comparison with an immediate value, we need to make the string constant data available to the native code. I think we'd need to add another parameter to the native `callFunction`, which would be a buffer of varlen constants.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/questdb/questdb?pullRequest=4177) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/questdb/questdb?pullRequest=4177) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/questdb/questdb?pullRequest=4177) it.</sub>

eugenels reviewed (commented):
Hello, and thank you for a good PR.

I've noticed the implementation of the 'scalar' version for JIT, and it looks good. I'm curious, do you have any benchmarks comparing the Java and scalar JIT? Is there any noticeable difference?

Additionally, have you considered an AVX version? I'm interested in your thoughts on its potential benefits/drawbacks and how you would go about implementing it.

mtopolnik commented:
> Hello, and thank you for a good PR.

Hello!
 
> I've noticed the implementation of the 'scalar' version for JIT, and it looks good. I'm curious, do you have any benchmarks comparing the Java and scalar JIT? Is there any noticeable difference?

I haven't run any benchmarks yet, and since this implementation uses data-dependent loads (indirection from `varLenIndex` to `column`), I'm not sure if there's going to be a big benefit. Since this lays out the groundwork, more improvements could be done in a followup PR.

> Additionally, have you considered an AVX version? I'm interested in your thoughts on its potential benefits/drawbacks and how you would go about implementing it.

I guess there's an opportunity to load a 32-byte vector which would contain four offsets. Then we can shift the vector right by 8 bytes and subtract, this would give us lengths of subsequent entries. If such a length is > 4, we know it's a non-null string. If it's = 4, we still have to follow the offset to see if the length is 0 or -1.

puzpuzpuz commented:
> If such a length is > 4, we know it's a non-null string. If it's = 4, we still have to follow the offset to see if the length is 0 or -1.

The same fast path could be introduced in scalar code. Would be nice to compare before/after on all non-null or mostly non-null column to understand if there is a speed-up.

mtopolnik commented on core/src/main/java/io/questdb/jit/CompiledFilterIRSerializer.java:
```diff
@@ -980,6 +987,9 @@ public int constantTypeCode() {
 
         public boolean hasMixedSizes() {
+            if (sizes[STR_INDEX] != 0) {
+                return true;
```
Although scalar mode is always suboptimal, at least we know this won't slow down previous queries, because a query involving a varlen column has so far been inelligible for JIT compilation.

mtopolnik commented:
All the tests are passing now, except failures that seem unrelated -- something with the formatting and rounding of floating-point numbers.

Next I can look into improving the null check JIT code.

mtopolnik commented:
I've looked into optimizing the scalar code. I wrote some initial machine code, but got stuck on sliding that into the existing framework. Current JIT code is organized around the assumption that processing each row means loading the value from the appropriate column. That's why it was easy to sneak into that the solution with indirect loading from a computed offset. However, when using the approach with subtracting subsequent offsets, I no longer have the length at a memory location, but in a register.

Since this is already quite an extensive PR, and there are still some issues with test failures, I propose we keep this PR focused on correctness, and isolate the performance-related changes into a separate PR.

mtopolnik commented:
These are the failing tests:

```LineTcpInsertOtherTypesTest.testInsertFloatTableExists
LineUdpInsertOtherTypesTest.testInsertFloatTableExists
InsertCastTest.testCastTimestampToFloatBind
UnionAllCastTest.testFloatInt
UnionAllCastTest.testFloatLong
SqlCompilerImplTest.testCastDateFloat
SqlCompilerImplTest.testCastIntFloat
SqlCompilerImplTest.testCastTimestampFloat
```

They fail on the `master` branch as well.

mtopolnik commented:
Thanks to @jerrinot I switched from testing with JDK 21 to 17, and that removed the failures. Apparently JDK 21 introduces different rules for FP rounding.

bluestreak01 commented:
thank you Marko, we are using JDK17 for this project. It might be easier!

mtopolnik commented:
I realized that the great asmjit API allows me to easily substitute a register for a memory location, so I went forward with the offset subtraction idea and added it to this PR.

mtopolnik commented:
Encouraged by the improved branchless solution with no data-dependent loads, I ran this simple benchmark:

```
create table x (str_value STRING);
insert into x select NULL from long_sequence(100000000);
insert into x values ('find me!');
x where str_value is not null
```

My timings were as follows:

```
master: 123 ms
jit-str: 63 ms
```

A 2x improvement is a good start!

mtopolnik commented:
I did another benchmark with a mostly non-null column:

```
create table x (str_value STRING);
insert into x select 'i am here!' from long_sequence(100000000);
insert into x values (null);
x where str_value is null
```
These were the timings:

```
master: 126 ms
jit-str: 88 ms
```

Timing on master is the same, as expected. But it's suprising that the timing with the JIT code is worse, since it's this case where it wouldn't have to do the data-dependent loads, and in the first case (mostly NULL colun) it was supposed to be doing them almost all the time.

mtopolnik commented:
Did the following benchmark now:

```
create table x (str_value STRING);
```

Insert 100,000,000 strings of variable length:

```
Random rnd = ThreadLocalRandom.current();
try (Sender sender = Sender.builder().address("localhost:9009").build()) {
    for (int i = 0; i < 100_000_000; i++) {
        sender.table("x")
                .stringColumn("str_value", "1234567".repeat(rnd.nextInt(20)))
                .at(0, ChronoUnit.NANOS);
    }
}
```

```
x where str_value is null (0 rows)
```

Timing:

```
jit-str: 378 ms
master: 250 ms
```

:(

puzpuzpuz commented:
> :(

Is it quick for you to try your initial version, without the offset check fast path?

mtopolnik commented:
I just did it, it performs slightly better at 370 ms. I also tried the latest version with just the `cmovz` instruction removed, and testing for zero value instead of -1: that made it dramatically faster, comparable to the initial measurement with short strings (60 ms).

mtopolnik commented:
The version with `jnz` instead of `cmovz` is better, 130 ms. It seems that `cmovz` loads that value whether it puts it into the register or not. I'll try to investigate this more.

mtopolnik commented:
Didn't have to look far...

>Specifically, CMOVcc loads data from its source operand into a temporary register unconditionally (regardless of the condition code and the status flags in the EFLAGS register).

puzpuzpuz commented:
> Didn't have to look far...

Nice finding!

mtopolnik commented:
Based on the finding, I'm reverting to the `jnz` approach.

mtopolnik commented:
Now i did a test with 100,000,000 non-null, non-empty strings plus a single null at the end: 60 ms. I get the exact same timing whether it's all short, same-size strings or longer, variable-size strings. Finally it all makes sense :) And it turns out that `jnz` doesn't really cost anything, since I also tested a (broken) variant without `jnz` and without `cmovz`, and it was the same 60 ms.

The case that gave me 130 ms (reported above) was with around 5% empty strings.

mtopolnik commented:
I ran a worst-case test as well, with a mix of 33% empty strings and 67% strings of two sizes (70 and 140 chars). 

In this case performance is worse: 270 ms this PR vs. 225 ms master. This was to be expected due to the large number of branch mispredictions.

mtopolnik commented:
I added AVX-2 code and removed the check that forces mixed mode whenever there's a string null check in the filter expression. I updated the PR description with benchmark results. I plan to rerun the GitHub action to rebuild the native code.

puzpuzpuz commented:
@eugenels could you do another review iteration? The PR seems ready to go.

eugenels commented:
please update the binaries
